File: src/main/java/dragonsaga/ui/components/FormHorizontalFloatScroll.java
package dragonsaga.ui.components;

import necesse.engine.localization.message.GameMessage;
import necesse.engine.localization.message.StaticMessage;
import necesse.gfx.forms.components.FormHorizontalScroll;

public class FormHorizontalFloatScroll extends FormHorizontalScroll<Float> {
    public FormHorizontalFloatScroll(int x, int y, int width, FormHorizontalScroll.DrawOption drawOption, GameMessage text, float value, float minValue, float maxValue) {
        super(x, y, width, drawOption, 0, new FormHorizontalScroll.ScrollElement[0]);
        this.set(text, value, minValue, maxValue);
    }

    public FormHorizontalFloatScroll(int x, int y, int width, FormHorizontalScroll.DrawOption drawOption, String text, float value, float minValue, float maxValue) {
        this(x, y, width, drawOption, new StaticMessage(text), value, minValue, maxValue);
    }

    public void set(GameMessage text, float value, float minValue, float maxValue) {

        int minValueInt = (int)Math.floor(minValue * 10);
        int maxValueInt = (int)Math.floor(maxValue * 10);

        FormHorizontalScroll.ScrollElement<Float>[] data = new FormHorizontalScroll.ScrollElement[maxValueInt - minValueInt + 1];

        for(int i = minValueInt; i <= maxValueInt; ++i) {
            int index = i - minValueInt;
            data[index] = new FormHorizontalScroll.ScrollElement(i / 10f, text);
        }

        this.setData(data);
        this.setElement(new FormHorizontalScroll.ScrollElement(value, text));
    }

    public void set(String text, float value, float minValue, float maxValue) {
        this.set(new StaticMessage(text), value, minValue, maxValue);
    }
}

File: src/main/kotlin/dragonsaga/ui/components/FormHorizontalFloatScroll.kt
package dragonsaga.ui.components

import necesse.engine.localization.message.GameMessage
import necesse.engine.localization.message.StaticMessage
import necesse.gfx.forms.components.FormHorizontalScroll

class FormHorizontalFloatScroll(x: Int, y: Int, width: Int, drawOption: DrawOption?, text: GameMessage?, value: Float, minValue: Float, maxValue: Float)
    : FormHorizontalScroll<Float>(x, y, width, drawOption, 0, *arrayOfNulls<ScrollElement<Float>>(0)) {

    init {
        this[text, value, minValue] = maxValue
    }

    constructor(x: Int, y: Int, width: Int, drawOption: DrawOption?, text: String?, value: Float, minValue: Float, maxValue: Float)
            : this(x, y, width, drawOption, StaticMessage(text), value, minValue, maxValue) {}

    operator fun set(text: GameMessage?, value: Float, minValue: Float, maxValue: Float) {
        val minValueInt = Math.floor((minValue * 10).toDouble()).toInt()
        val maxValueInt = Math.floor((maxValue * 10).toDouble()).toInt()
        val data: Array<ScrollElement<Float>?> = arrayOfNulls(maxValueInt - minValueInt + 1)

        for (i in minValueInt..maxValueInt) {
            val index = i - minValueInt
            data[index] = ScrollElement<Float>(i / 10f, text)
        }
        setData(data)
        setElement(ScrollElement(value, text))
    }

    operator fun set(text: String?, value: Float, minValue: Float, maxValue: Float) {
        this[StaticMessage(text), value, minValue] = maxValue
    }
}